src/tree/node.rs

```diff
@@ -0,0 +1,384 @@
+use std::cmp::Ordering;
+use std::cmp::Ordering::*;
+use std::mem::{self, replace, swap};
+
+use compare::Compare;
+
+// Nodes keep track of their level in the tree, starting at 1 in the
+// leaves and with a red child sharing the level of the parent.
+#[derive(Clone)]
+pub struct Node<K, V> {
+    key: K,
+    value: V,
+    left: Option<Box<Node<K, V>>>,
+    right: Option<Box<Node<K, V>>>,
+    level: usize,
+}
+
+impl<K, V> Node<K, V> {
+    pub fn key(&self) -> &K { &self.key }
+
+    pub fn value(&self) -> &V { &self.value }
+
+    pub fn value_mut(&mut self) -> &mut V { &mut self.value }
+
+    pub fn key_value_mut(&mut self) -> (&K, &mut V) { (&self.key, &mut self.value) }
+}
+
+impl<K, V> Node<K, V> {
+    #[inline]
+    fn new(key: K, value: V) -> Node<K, V> {
+        Node {key: key, value: value, left: None, right: None, level: 1 }
+    }
+
+    // Remove left horizontal link by rotating right
+    fn skew(&mut self) {
+        if self.left.as_ref().map_or(false, |x| x.level == self.level) {
```

Comment:
Gonna just assume this code is basically right. Not super in the mood to audit red-black-tree logic >.>